### PR TITLE
feat: Including use of GeoModel toolkit

### PR DESCRIPTION
**Pull request by noemina**
Including configuration to use GeoModel plugins in examples.
This MR also includes updates to the documentation.

### CODE REVIEW

**Comment by asalzburger** [comment]
Is this WIP because we expect still changes? Otherwise, let's just un-WIP it.

File: Examples/Algorithms/Geant4/src/GdmlDetectorConstruction.cpp [hunk 1]
```diff
@@ -1,23 +0,0 @@
...
-
-GdmlDetectorConstruction::GdmlDetectorConstruction(std::string path)
-    : G4VUserDetectorConstruction(), m_path(std::move(path)) {}
-
-G4VPhysicalVolume* GdmlDetectorConstruction::Construct() {
-  G4GDMLParser parser;
-  // TODO how to handle errors
-  parser.Read(m_path);
-  return parser.GetWorldVolume();
-}
```
Comment (paulgessinger) [inline Examples/Algorithms/Geant4/src/GdmlDetectorConstruction.cpp:23]: So this is replaced/superceded by GDML construction in GeoModelG4?

File: Examples/Algorithms/Geant4GeoModel/include/ActsExamples/Geant4GeoModel/GdmlDetectorConstruction.hpp [hunk 2]
```diff
@@ -24,6 +26,9 @@ class GdmlDetectorConstruction final : public G4VUserDetectorConstruction {
 
  private:
   std::string m_path;
+
+  /// Construct a GeoModel world volume
+  GeoPhysVol* CreateTheWorld();
```
Comment (paulgessinger) [inline Examples/Algorithms/Geant4GeoModel/include/ActsExamples/Geant4GeoModel/GdmlDetectorConstruction.hpp:31]:
```suggestion
  GeoPhysVol* CreateWorld();
```

File: Examples/Algorithms/Geant4GeoModel/src/GdmlDetectorConstruction.cpp [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+G4VPhysicalVolume* GdmlDetectorConstruction::Construct() {
+  // Load and build the geometry from GeoModel
+  GeoPhysVol* world = CreateTheWorld();
+  GeoGeometryPluginLoader loader;
+  GeoVGeometryPlugin* factory = loader.load(m_path);
+  factory->create(world);
+
+  // build the Geant4 geometry and get an hanlde to the world' volume
+  ExtParameterisedVolumeBuilder* builder =
+      new ExtParameterisedVolumeBuilder("ATLAS");
```
Comment (paulgessinger) [inline Examples/Algorithms/Geant4GeoModel/src/GdmlDetectorConstruction.cpp:38]: Are we assuming GeoModel == ATLAS?

File: Examples/Algorithms/Geant4GeoModel/include/ActsExamples/Geant4GeoModel/GdmlDetectorConstruction.hpp (hunk 2 above)
Comment (paulgessinger) [reply Examples/Algorithms/Geant4GeoModel/include/ActsExamples/Geant4GeoModel/GdmlDetectorConstruction.hpp:31]: Could this be `std::unique_ptr`, so it's clear where the ownership is? Does GeoModel retain ownership over the top volume?

File: Examples/Algorithms/Geant4GeoModel/src/GdmlDetectorConstruction.cpp (hunk 3 above)
Comment (paulgessinger) [reply Examples/Algorithms/Geant4GeoModel/src/GdmlDetectorConstruction.cpp:38]: I think this is a memory leak. Does it have to be allocated on the heap or could it just be a stack object?

File: Examples/Algorithms/Geant4GeoModel/src/GdmlDetectorConstruction.cpp [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+  G4LogicalVolume* envelope = builder->Build(world);
+  G4VPhysicalVolume* physWorld = new G4PVPlacement(
+      0, G4ThreeVector(), envelope, envelope->GetName(), 0, false, 0, false);
+
+  G4VPhysicalVolume* fWorld = physWorld;
```
Comment (paulgessinger) [inline Examples/Algorithms/Geant4GeoModel/src/GdmlDetectorConstruction.cpp:43]: The alias is redundant, right?

File: Examples/Algorithms/Geant4GeoModel/src/GdmlDetectorConstruction.cpp [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+  // -------------------------------------//
+  GeoMaterial* Air = new GeoMaterial(
+      "Air", 1.290 * SYSTEM_OF_UNITS::mg / SYSTEM_OF_UNITS::cm3);
+  GeoElement* Oxigen = new GeoElement(
+      "Oxygen", "O", 8.0, 16.0 * SYSTEM_OF_UNITS::g / SYSTEM_OF_UNITS::mole);
+  GeoElement* Nitrogen =
+      new GeoElement("Nitrogen", "N", 7.,
+                     14.0067 * SYSTEM_OF_UNITS::g / SYSTEM_OF_UNITS::mole);
+
+  Air->add(Nitrogen, .8);
```
Comment (paulgessinger) [inline Examples/Algorithms/Geant4GeoModel/src/GdmlDetectorConstruction.cpp:62]: Does `GeoMaterial` take ownership of `Nitrogen` here?

File: Examples/Algorithms/Geant4GeoModel/src/GdmlDetectorConstruction.cpp [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+
+  return fWorld;
+}
+
+GeoPhysVol* GdmlDetectorConstruction::CreateTheWorld() {
+  GeoPhysVol* world = nullptr;
+  // Setup the 'World' volume from which everything else will be suspended
+  // Get the materials that we shall use.
+  // -------------------------------------//
+  GeoMaterial* Air = new GeoMaterial(
```
Comment (paulgessinger) [inline Examples/Algorithms/Geant4GeoModel/src/GdmlDetectorConstruction.cpp:54]: We usually don't capitalize variable names:

```suggestion
  GeoMaterial* air = new GeoMaterial(
```

There are a few variables below which are also affected.

File: Examples/Algorithms/Geant4GeoModel/src/GdmlDetectorConstruction.cpp [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+      "Air", 1.290 * SYSTEM_OF_UNITS::mg / SYSTEM_OF_UNITS::cm3);
+  GeoElement* Oxigen = new GeoElement(
```
Comment (paulgessinger) [inline Examples/Algorithms/Geant4GeoModel/src/GdmlDetectorConstruction.cpp:56]:
```suggestion
  GeoElement* oxygen = new GeoElement(
```

File: CMakeLists.txt [hunk 8]
```diff
@@ -222,6 +223,10 @@ endif()
 if(ACTS_BUILD_EXAMPLES_GEANT4)
   find_package(Geant4 REQUIRED CONFIG COMPONENTS gdml)
 endif()
+if(ACTS_BUILD_EXAMPLES_GEOMODEL)
+  find_package( GeoModelCore PATHS ${GeoModel_INSTALL}  REQUIRED ${GeoModel_VERSION} )
+  find_package( GeoModelG4 PATHS ${GeoModel_INSTALL} REQUIRED ${GeoModel_VERSION} )
```
Comment (paulgessinger) [inline CMakeLists.txt:228]: The question is, do we bundle GeoModel? I think it's small enough that it could be viable, but not sure if it's necessary.

**Review by paulgessinger — COMMENTED** [review]
This looks pretty thorough. I added a few comments and questions.

The CMake changes seem fine to me, if our intention is indeed that all GDML input now goes through GeoModel. Is that the intention? Can GeoModel handle any GDML input?

File: Examples/Algorithms/Geant4/src/GdmlDetectorConstruction.cpp (hunk 1 above)
Comment (noemina) [reply Examples/Algorithms/Geant4/src/GdmlDetectorConstruction.cpp:23]: Yes, @FabianKlimpel and I have seen issues loading directly the `gdml` via the `G4GDMLParser`, so we have decided to include `GeoModel` to do it. More general, plugins available in `GeoModel` allow to build geometries from several sourced, e.g. `*.db` or `*.xml`, sharing the same files ATLAS uses.

File: Examples/Algorithms/Geant4GeoModel/include/ActsExamples/Geant4GeoModel/GdmlDetectorConstruction.hpp (hunk 2 above)
Comment (noemina) [reply Examples/Algorithms/Geant4GeoModel/include/ActsExamples/Geant4GeoModel/GdmlDetectorConstruction.hpp:31]: This is a good question. I will have a look at this and the additional points you have raised below.

File: Examples/Algorithms/Geant4/src/GdmlDetectorConstruction.cpp (hunk 1 above)
Comment (paulgessinger) [reply Examples/Algorithms/Geant4/src/GdmlDetectorConstruction.cpp:23]: Ok, great. If we don't expect a use case for importing GDML bypassing GeoModel, I think this is fine then.

**Comment by asalzburger** [comment]
Hi, I would be in favour of having a:
 * a GDML based input that is called `GdmlDetectorConstruction` that feeds directly into Geant4
 * a GDML-GeoModel one that can handle ATLAS ITk e.g. and we call it `GdmlGeoModelDetectorConstruction`

**Comment by asalzburger** [comment]
Also, from my last information I gather that this only works with an older version of GeoModel, so we may need to put that into the requrirements.

**Comment by noemina** [comment]
Hello, 
Some time ago I reported the problem seed when loading the `gdml` (which was the reason I started including the GeoModel toolkit to load the `gdml` via a different plugin rather than using the `gdml` reader from Geant4) to the GeoModel developers, as the same issue was appearing there: https://gitlab.cern.ch/GeoModelDev/GeoModel/-/issues/21 
It was found that there was a problem in the way the `gdml` was produced in the ATLAS athena release and has been fixed as reported in the issue above.
I will close this PR as this is not needed (at least now) any more in order to avoid complication in maintainability.
Cheers,
Noemi

**Pull request closed without merging**